apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,702 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Compute.v1.Data;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Logging;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Apis.Compute.v1.Data.Metadata;
+
+namespace Google.Cloud.Metadata.V1
+{
+    public sealed class MetadataClientImpl : MetadataClient
+    {
+        private const string DefaultMetadataHost = "metadata";
+        private const string EmulatorEnvironmentVariable = "METADATA_EMULATOR_HOST";
+        private const string MetadataFlavor = "Metadata-Flavor";
+        private const string GoogleMetadataHeader = "Google";
+
+        // TODO: Should this be created with a MetadataClient instead?
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<MetadataClientImpl>();
+
+        private readonly string metadataHost;
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        public MetadataClientImpl(IHttpClientFactory httpClientFactory)
+        {
+            HttpClient = (httpClientFactory ?? new HttpClientFactory()).CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        private static MaintenanceStatus GetMaintenanceStatus(string result)
+        {
+            switch (result)
+            {
+                case "NONE": return MaintenanceStatus.None;
+                case "MIGRATE_ON_HOST_MAINTENANCE": return MaintenanceStatus.Migrate;
+                default: return MaintenanceStatus.Unknown;
+            }
+        }
+
+        private string GetMetadataHost()
```

Comment:
We're not quite sure what we're doing about emulator discovery for the rest of the APIs yet.  We'll want to end up with them being consistent, of course, but we can start off with this.